source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1132,7 +1133,120 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::drainConnP
 ClusterUpdateCallbacksHandlePtr
 ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks& cb) {
   ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
-  return std::make_unique<ClusterUpdateCallbacksHandleImpl>(cb, cluster_manager.update_callbacks_);
+  return cluster_manager.addClusterUpdateCallbacks(cb);
+}
+
+OdCdsApiHandlePtr
+ClusterManagerImpl::allocateOdCdsApi(const envoy::config::core::v3::ConfigSource& odcds_config,
+                                     OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
+                                     ProtobufMessage::ValidationVisitor& validation_visitor) {
+  // TODO(krnowak): Instead of creating a new handle every time, store the handles internally and
```

Comment:
> I agree with this TODO.
> One thing to note is that at the moment a connection/stream will be created for every instance of ODCDS config (even if the configuration is the same). This is even worse if EDS is used, as it will double the EDS subscriptions.

Unless you use ADS, that is.

> Another thing to take into consideration is that CDS isn't warmed during LDS/VHDS/RDS config processing, so a request will be sent when processing each occurrence of an ODCDS config.

I'm not sure I understand this one. I know that there is cluster warming up, but CDS warming up?

There is also another thing to consider - deduplication also means fewer connections used, which means that one connection will hold the subscription to resources requested on demand until last user of this connection goes away. Which means that instead of having say 5 connections having a subscription to 3 resources each, we would have 1 connection with subscription to 15 resources.

Reply:
> I probably should've written my comment somewhere else, as it is somewhat orthogonal. It is referring to the number of requests that will be sent and not the number of connections/handlers.
> 
> Re ADS, note that at the moment warming resources is being done by pausing the muxer from sending certain types of requests (see
> 
> https://github.com/envoyproxy/envoy/blob/3b8296d70600e5280e4d3ece5addacec0b74aa1c/source/common/upstream/cds_api_helper.cc#L20-L27
> for example of CDS->EDS).
> 
> Re LDS/VHDS/RDS, I've looked at the integration tests at https://github.com/kinvolk/envoy/blob/krnowak/odcds/test/integration/odcds_integration_test.cc and it doesn't seem that there is warming, so just wanted to mention that.
> 
> No warming implies that multiple requests will be sent, even to the same ODCDS endpoint.

I must admit that I still don't get the idea behind warming of xDSes. Is there some documentation about it, or code, so I could read about it and wrap my head around the concept?

I can only say that ODCDS uses the CDS API helper you linked above, so it also should be doing the pausing the muxer for ADS.

I was also confused about "a request will be sent when processing each occurrence of an ODCDS config". What request are we talking here about?